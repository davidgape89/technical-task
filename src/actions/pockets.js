export const exchangeCurrency = (fromCurrency, fromAmount, toCurrency, toAmount) => ({
  type: 'EXCHANGE_CURRENCY',
  fromCurrency,
  fromAmount,
  toCurrency,
  toAmount
});