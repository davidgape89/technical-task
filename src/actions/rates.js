import axios from 'axios';

// Export for testing
export const ratesUrl = `https://api.exchangeratesapi.io/latest`;

export const setBaseCurrency = (base) => {
  return (dispatch) => {
    dispatch({type: 'SET_BASE', base});
    dispatch(requestRates());
  }
}

export const requestRates = () => {
  return (dispatch, getState) => {
    const {rates: {base}} = getState();
    
    return axios.get(`${ratesUrl}?base=${base}`)
      .then((response) => {
        let rates = response.data.rates;
        rates[base] = 1;
        dispatch(setRates(rates));
      })
      .catch((error) => 
        console.error('There was a problem fetching the rates'));
  }
}

export const setRates = (rates) => ({
  type: 'SET_RATES',
  rates
});