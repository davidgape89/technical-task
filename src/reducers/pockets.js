export const defaultState = {
  'EUR': 20.000000,
  'USD': 40.000000,
  'GBP': 80.000000
};

export const pocketsReducer = (state = defaultState, action) => {
  switch(action.type) {
    case 'SET_POCKET': 
      return {...state, [action.currency]: action.amount};
    case 'EXCHANGE_CURRENCY':
      const fromCurrency = action.fromCurrency;
      const toCurrency = action.toCurrency;
      const fromAmount = action.fromAmount;
      const toAmount = action.toAmount;
      return {
        ...state,
        [fromCurrency]: state[fromCurrency] - fromAmount,
        [toCurrency]: state[toCurrency] + toAmount
      };
    default:
      return state;
  }
}