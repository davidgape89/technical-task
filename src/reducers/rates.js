export const defaultState = {
  base: 'EUR',
  rates: {}
}

export const ratesReducer = (state = defaultState, action) => {
  switch(action.type) {
    case 'SET_BASE':
      return {...state, base: action.base};
    case 'SET_RATES':
      return {...state, rates: action.rates};
    default:
      return state;
  }
} 