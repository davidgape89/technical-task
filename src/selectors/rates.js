export const exchangeToValue = (rates, fromCurrency, toCurrency, amount) => amount / rates[fromCurrency] * rates[toCurrency];

export const exchangeFromValue = (rates, fromCurrency, toCurrency, amount) => amount / rates[toCurrency] * rates[fromCurrency];
