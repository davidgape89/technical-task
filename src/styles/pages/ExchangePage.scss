@import '../variables';

h1 {
  color: $blue-600;
  padding-bottom: 16px;
}

.exchange-page {
  > * {
    margin-bottom: 20px;
  }
}

.exchange-page__from-row,
.exchange-page__to-row {
  display: flex;
}

.exchange-page__from-select,
.exchange-page__to-select,
.exchange-page__switcher {
  flex: 1;
  margin-right: 30px;
}

.exchange-page__from-input,
.exchange-page__to-input,
.exchange-page__currency-info {
  flex: 2;
}

.exchange-page__utils {
  display: flex;
  align-items: center;
}

.exchange-page__switcher,
.exchange-page__currency-info {
  display: flex;
  justify-content: center;
}

.exchange-page__currency-info {
  line-height: 25px;

  .material-icons {
    margin-right: 8px;
  }
}



