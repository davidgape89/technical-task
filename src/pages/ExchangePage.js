import React, { useState } from 'react';
import propTypes from 'prop-types';
import { connect } from 'react-redux';
import Button from '@material-ui/core/Button';
import Fab from '@material-ui/core/Fab';
import Icon from '@material-ui/core/Icon';

import exchangeCurrencyFunc from '../actions/pockets';
import CurrencySelect from '../components/CurrencySelect';
import CurrencyInput from '../components/CurrencyInput';
import CurrencyExchangePanel from '../components/CurrencyExchangePanel';
import { exchangeToValue, exchangeFromValue } from '../selectors/rates';
import { CURRENCY_SYMBOLS } from '../constants';

export const ExchangePage = ({ pockets, rates, exchangeCurrency }) => {
  const pocketCurrencies = Object.keys(pockets);
  const [state, setState] = useState({
    fromCurrency: pocketCurrencies[0],
    toCurrency: pocketCurrencies[1],
    toValue: '',
    fromValue: '',
    isDisabled: true,
  });

  const onFromValueChange = (value) => {
    setState((prevState) => {
      const newToValue = value
        ? exchangeToValue(
          rates,
          prevState.fromCurrency,
          prevState.toCurrency,
          parseFloat(value),
        ).toString()
        : '';
      return {
        ...prevState,
        fromValue: value,
        toValue: newToValue,
      };
    });
  };

  const onToValueChange = (value) => {
    setState((prevState) => {
      const newFromValue = value
        ? exchangeFromValue(
          rates,
          prevState.fromCurrency,
          prevState.toCurrency,
          parseFloat(value),
        ).toString()
        : '';
      return ({
        ...prevState,
        toValue: value,
        fromValue: newFromValue,
      });
    });
  };

  const onFromCurrencyChange = (value) => {
    setState(prevState => ({
      ...prevState,
      toCurrency: prevState.toCurrency === value
        ? prevState.fromCurrency : prevState.toCurrency,
      fromCurrency: value,
    }));

    onFromValueChange(state.fromValue);
  };

  const onToCurrencyChange = (value) => {
    setState(prevState => ({
      ...prevState,
      fromCurrency: prevState.fromCurrency === value
        ? prevState.toCurrency : prevState.fromCurrency,
      toCurrency: value,
    }));

    onToValueChange(state.toValue);
  };

  const switchCurrencies = () => {
    setState(prevState => ({
      ...prevState,
      toCurrency: prevState.fromCurrency,
      fromCurrency: prevState.toCurrency,
      toValue: prevState.fromValue,
      fromValue: prevState.toValue,
    }));
  };

  const isButtonDisabled = () => {
    const floatValue = parseFloat(state.fromValue);
    const isDisabled = !floatValue || floatValue > pockets[state.fromCurrency];
    return isDisabled;
  };

  const onExchangeClick = () => {
    const toValue = parseFloat(state.toValue);
    const fromValue = parseFloat(state.fromValue);

    exchangeCurrency(
      state.fromCurrency,
      fromValue,
      state.toCurrency,
      toValue,
    );

    isButtonDisabled();
  };

  return (
    <div className="exchange-page">
      <h1>Currency Exchange</h1>
      <div className="exchange-page__from-row">
        <div className="exchange-page__from-select">
          <CurrencySelect
            currencies={pockets}
            onChange={onFromCurrencyChange}
            value={state.fromCurrency}
          />
        </div>
        <div className="exchange-page__from-input">
          <CurrencyInput
            value={state.fromValue}
            placeholder="0"
            label="Amount to Exchange"
            required
            symbol={CURRENCY_SYMBOLS[state.fromCurrency]}
            onChange={onFromValueChange}
          />
        </div>
      </div>
      <div className="exchange-page__utils">
        <div className="exchange-page__switcher">
          <Fab color="primary" onClick={switchCurrencies}>
            <Icon>cached</Icon>
          </Fab>
        </div>
        <div className="exchange-page__currency-info">
          <Icon>show_chart</Icon>
          <CurrencyExchangePanel
            rates={rates}
            fromCurrency={state.fromCurrency}
            toCurrency={state.toCurrency}
          />
        </div>
      </div>
      <div className="exchange-page__to-row">
        <div className="exchange-page__to-select">
          <CurrencySelect
            currencies={pockets}
            onChange={onToCurrencyChange}
            value={state.toCurrency}
          />
        </div>
        <div className="exchange-page__to-input">
          <CurrencyInput
            value={state.toValue}
            placeholder="0"
            label="Exchanged Amount"
            symbol={CURRENCY_SYMBOLS[state.toCurrency]}
            onChange={onToValueChange}
          />
        </div>
      </div>
      <Button
        fullWidth
        disabled={isButtonDisabled()}
        variant="contained"
        color="primary"
        onClick={onExchangeClick}
      >
        Exchange
      </Button>
    </div>
  );
};

ExchangePage.propTypes = {
  pockets: propTypes.objectOf(propTypes.number).isRequired,
  rates: propTypes.objectOf(propTypes.number).isRequired,
  exchangeCurrency: propTypes.func.isRequired,
};

const mapStateToProps = state => ({
  pockets: state.pockets,
  rates: state.rates.rates,
});

const mapDispatchToProps = {
  exchangeCurrency: exchangeCurrencyFunc,
};

export default connect(mapStateToProps, mapDispatchToProps)(ExchangePage);
