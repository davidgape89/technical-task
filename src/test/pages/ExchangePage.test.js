import React from 'react';
import {shallow} from 'enzyme';
import Fab from '@material-ui/core/Fab';
import Button from '@material-ui/core/Button';

import pockets from '../fixtures/pockets';
import rates from '../fixtures/rates';
import {ExchangePage} from '../../pages/ExchangePage';

describe('CurrencyExchangePanel - ', () => {
  let wrapper, exchangeCurrency;
  exchangeCurrency = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <ExchangePage pockets={pockets}
                    rates={rates.rates}
                    exchangeCurrency={exchangeCurrency}/>
    );
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('onFromCurrencyChange -', () => {
    it('defaults correctly', () => {
      const fromCurrency = wrapper.find(
        '.exchange-page__from-row CurrencySelect').prop('value');
      const toCurrency = wrapper.find(
        '.exchange-page__to-row CurrencySelect').prop('value');

      expect(fromCurrency).toEqual('EUR');
      expect(toCurrency).toEqual('USD');
    });

    it('changes correctly', () => {
      wrapper.find(
        '.exchange-page__from-row CurrencySelect').prop('onChange')('GBP');
      const fromCurrency = wrapper.find(
        '.exchange-page__from-row CurrencySelect').prop('value');

      expect(fromCurrency).toEqual('GBP');
    });

    it('interchanges the currencies correctly', () => {
      wrapper.find(
        '.exchange-page__from-row CurrencySelect').prop('onChange')('USD');
      const fromCurrency = wrapper.find(
        '.exchange-page__from-row CurrencySelect').prop('value');
      const toCurrency = wrapper.find(
        '.exchange-page__to-row CurrencySelect').prop('value');

      expect(fromCurrency).toEqual('USD');
      expect(toCurrency).toEqual('EUR');
    });
  });

  describe('onToCurrencyChange -', () => {
    it('defaults correctly', () => {
      const fromCurrency = wrapper.find(
        '.exchange-page__from-row CurrencySelect').prop('value');
      const toCurrency = wrapper.find(
        '.exchange-page__to-row CurrencySelect').prop('value');

      expect(fromCurrency).toEqual('EUR');
      expect(toCurrency).toEqual('USD');
    });

    it('changes correctly', () => {
      wrapper.find(
        '.exchange-page__to-row CurrencySelect').prop('onChange')('GBP');
      const toCurrency = wrapper.find(
        '.exchange-page__to-row CurrencySelect').prop('value');

      expect(toCurrency).toEqual('GBP');
    });

    it('interchanges the currencies correctly', () => {
      wrapper.find(
        '.exchange-page__to-row CurrencySelect').prop('onChange')('EUR');
      const fromCurrency = wrapper.find(
        '.exchange-page__from-row CurrencySelect').prop('value');
      const toCurrency = wrapper.find(
        '.exchange-page__to-row CurrencySelect').prop('value');

      expect(fromCurrency).toEqual('USD');
      expect(toCurrency).toEqual('EUR');
    });
  });

  describe('onFromValueChange -', () => {
    it('defaults correctly', () => {
      const fromValue = wrapper.find(
        '.exchange-page__from-row CurrencyInput').prop('value');
      
      expect(fromValue).toEqual('');
    });

    it('changes correctly', () => {
      wrapper.find(
        '.exchange-page__from-row CurrencyInput').prop('onChange')('1');
      const fromValue = wrapper.find(
        '.exchange-page__from-row CurrencyInput').prop('value');
      const toValue = wrapper.find(
        '.exchange-page__to-row CurrencyInput').prop('value');

      expect(fromValue).toBe('1');
      expect(toValue).toBe('1.1354');
    });

    it('goes back to empty correctly', () => {
      wrapper.find(
        '.exchange-page__from-row CurrencyInput').prop('onChange')('');
      const fromValue = wrapper.find(
        '.exchange-page__from-row CurrencyInput').prop('value');
      const toValue = wrapper.find(
        '.exchange-page__to-row CurrencyInput').prop('value');

      expect(fromValue).toBe('');
      expect(toValue).toBe('');
    });
  });

  describe('onToValueChange -', () => {
    it('defaults correctly', () => {
      const toValue = wrapper.find(
        '.exchange-page__to-row CurrencyInput').prop('value');
      
      expect(toValue).toEqual('');
    });

    it('changes correctly', () => {
      wrapper.find(
        '.exchange-page__to-row CurrencyInput').prop('onChange')('1.1354');
      const fromValue = wrapper.find(
        '.exchange-page__from-row CurrencyInput').prop('value');
      const toValue = wrapper.find(
        '.exchange-page__to-row CurrencyInput').prop('value');

      expect(fromValue).toBe('1');
      expect(toValue).toBe('1.1354');
    });

    it('goes back to empty correctly', () => {
      wrapper.find(
        '.exchange-page__to-row CurrencyInput').prop('onChange')('');
      const fromValue = wrapper.find(
        '.exchange-page__from-row CurrencyInput').prop('value');
      const toValue = wrapper.find(
        '.exchange-page__to-row CurrencyInput').prop('value');

      expect(fromValue).toBe('');
      expect(toValue).toBe('');
    });
  });

  describe('switchCurrencies -', () => {
    it('exchanges inputs correctly', () => {
      wrapper.find(
        '.exchange-page__from-row CurrencyInput').prop('onChange')('1');
      wrapper.find(Fab).simulate('click');
  
      const fromCurrency = wrapper.find(
        '.exchange-page__from-row CurrencySelect').prop('value');
      const toCurrency = wrapper.find(
        '.exchange-page__to-row CurrencySelect').prop('value');
      const fromValue = wrapper.find(
        '.exchange-page__from-row CurrencyInput').prop('value');
      const toValue = wrapper.find(
        '.exchange-page__to-row CurrencyInput').prop('value');
  
      expect(fromCurrency).toEqual('USD');
      expect(toCurrency).toEqual('EUR');
      expect(fromValue).toEqual('1.1354');
      expect(toValue).toEqual('1');
    });
  });

  describe('isButtonDisabled -', () => {
    it('is disabled by default', () => {
      expect(wrapper.find(Button).prop('disabled')).toBe(true);
    });

    it('is enabled with valid attributes', () => {
      wrapper.find(
        '.exchange-page__from-row CurrencyInput').prop('onChange')('1');

        expect(wrapper.find(Button).prop('disabled')).toBe(false);
    });

    it('is disabled when the amount exeeds the funds', () => {
      wrapper.find(
        '.exchange-page__from-row CurrencyInput').prop('onChange')('50');

      expect(wrapper.find(Button).prop('disabled')).toBe(true);
    });
  });

  describe('onExchangeClick -', () => {
    it('exchanges the amount when clicked', () => {
      wrapper.find(
        '.exchange-page__from-row CurrencyInput').prop('onChange')('1');
      wrapper.find(Button).simulate('click');
      
      expect(exchangeCurrency).toHaveBeenCalled();
    });
  });
});