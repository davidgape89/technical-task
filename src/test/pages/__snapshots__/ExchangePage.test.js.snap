// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyExchangePanel -  renders correctly 1`] = `
<div
  className="exchange-page"
>
  <h1>
    Currency Exchange
  </h1>
  <div
    className="exchange-page__from-row"
  >
    <div
      className="exchange-page__from-select"
    >
      <CurrencySelect
        currencies={
          Object {
            "EUR": 20,
            "GBP": 40,
            "USD": 30,
          }
        }
        isBalanceRed={false}
        onChange={[Function]}
        value="EUR"
      />
    </div>
    <div
      className="exchange-page__from-input"
    >
      <CurrencyInput
        label="Amount to Exchange"
        onChange={[Function]}
        placeholder="0"
        required={true}
        symbol="â‚¬"
        value=""
      />
    </div>
  </div>
  <div
    className="exchange-page__utils"
  >
    <div
      className="exchange-page__switcher"
    >
      <WithStyles(Fab)
        color="primary"
        onClick={[Function]}
      >
        <WithStyles(Icon)>
          cached
        </WithStyles(Icon)>
      </WithStyles(Fab)>
    </div>
    <div
      className="exchange-page__currency-info"
    >
      <WithStyles(Icon)>
        show_chart
      </WithStyles(Icon)>
      <CurrencyExchangePanel
        fromCurrency="EUR"
        rates={
          Object {
            "AUD": 1.5999,
            "BGN": 1.9558,
            "BRL": 4.3012,
            "CAD": 1.5125,
            "CHF": 1.1338,
            "CNY": 7.6014,
            "CZK": 25.646,
            "DKK": 7.4624,
            "EUR": 1,
            "GBP": 0.8628,
            "HKD": 8.9128,
            "HRK": 7.4178,
            "HUF": 313.29,
            "IDR": 16082.94,
            "ILS": 4.0974,
            "INR": 78.136,
            "ISK": 132.9,
            "JPY": 126.63,
            "KRW": 1281.96,
            "MXN": 21.5168,
            "MYR": 4.6109,
            "NOK": 9.6915,
            "NZD": 1.6567,
            "PHP": 60.042,
            "PLN": 4.2834,
            "RON": 4.7615,
            "RUB": 73.004,
            "SEK": 10.431,
            "SGD": 1.5335,
            "THB": 36.032,
            "TRY": 6.2152,
            "USD": 1.1354,
            "ZAR": 16.3783,
          }
        }
        toCurrency="USD"
      />
    </div>
  </div>
  <div
    className="exchange-page__to-row"
  >
    <div
      className="exchange-page__to-select"
    >
      <CurrencySelect
        currencies={
          Object {
            "EUR": 20,
            "GBP": 40,
            "USD": 30,
          }
        }
        isBalanceRed={false}
        onChange={[Function]}
        value="USD"
      />
    </div>
    <div
      className="exchange-page__to-input"
    >
      <CurrencyInput
        label="Exchanged Amount"
        onChange={[Function]}
        placeholder="0"
        symbol="$"
        value=""
      />
    </div>
  </div>
  <WithStyles(Button)
    color="primary"
    disabled={true}
    fullWidth={true}
    onClick={[Function]}
    variant="contained"
  >
    Exchange
  </WithStyles(Button)>
</div>
`;
